assets/app/view/game/part/town_location.rb

```diff
@@ -0,0 +1,250 @@
+        # positioning the town rectangle; this value will be added/subtracted to
+        # (edge_a * 60)
+        POSITIONAL_ANGLE = {
+          sharp: 12.12,
+          gentle: 6.11,
+        }.freeze
+
+        # Returns the two edges plus min edge so that a is the edge to render close to, and
+        # (a - b).abs is 1, 2 or 3.
+        def normalized_edges(edge_a, exits)
+          @@normalized_edges ||= {}
+          @@normalized_edges[[edge_a, exits]] ||= if edge_a && exits.size == 2
+                                                    edge_b = exits.find { |e| e != edge_a }
+                                                    edges = [edge_a, edge_b]
+                                                    edges[edges.index(edges.min)] += 6 if (edge_a - edge_b).abs > 3
+                                                    edges.append(edges.min)
+                                                  else
+                                                    [edge_a, nil]
+                                                  end
+        end
+
+        def min_edge(edges)
+          @@min_edge ||= {}
+          @@min_edge[edges] ||= edges.min
+        end
+
+        # Returns a symbol of :straight, :gentle or :sharp based on the absolute
+        # difference of the normalized edges
+        def town_track_type(edges)
+          @@town_track_type ||= {}
+          @@town_track_type[edges] ||=
+            begin
+              edge_a, edge_b = edges
+              [nil, :sharp, :gentle, :straight][(edge_a - edge_b).abs]
+            end
+        end
+
+        # Returns a symbol of :left, :right, or :straight based on the direction
+        # the track curves starting from edge_a
+        def town_track_direction(edges)
+          @@town_track_direction ||= {}
+          @@town_track_direction[edges] ||=
+            begin
+              edge_a, edge_b = edges
+
+              if (edge_a - edge_b).abs == 3
+                :straight
+              elsif edge_a > edge_b
+                :right
+              elsif edge_a < edge_b
+                :left
+              end
+            end
+        end
+
+        # Returns an array of options for track positions
+        def town_track_location(tile, town, edges)
+          edge_a, = edges
+
+          loc = if town.exits.size == 2 && tile.stops.one?
+                  SINGLE_STOP_TWO_EXIT_REGIONS[town_track_type(edges)][min_edge(edges)]
+                elsif town.exits.size == 2
+                  MULTIPLE_STOP_TWO_EXIT_REGIONS[town_track_type(edges)][edge_a % 6]
+                elsif edge_a
+                  EDGE_TOWN_REGIONS[edge_a]
+                else
+                  CENTER
+                end
+          [loc]
+        end
+
+        # Returns an array of rotation options for the town rectangle that
+        # corresponds to the positions given from track_location
+        def town_rotation_angles(tile, town, edges)
+          edge_a, = edges
```

Comment:
should this be cacheD?

Reply:
Hmmm.  It will only be called once with a given town.